sdks/python/apache_beam/io/filesystemio.py

```diff
@@ -0,0 +1,492 @@
+#
+# Licensed to the Apache Software Foundation (ASF) under one or more
+# contributor license agreements.  See the NOTICE file distributed with
+# this work for additional information regarding copyright ownership.
+# The ASF licenses this file to You under the Apache License, Version 2.0
+# (the "License"); you may not use this file except in compliance with
+# the License.  You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+"""Utilities for ``FileSystem`` implementations."""
+
+import abc
+import cStringIO
+import logging
+import multiprocessing
+import os
+import traceback
+
+__all__ = ['Downloader', 'Uploader', 'BufferedReader', 'BufferedWriter']
+
+
+class Downloader(object):
+  """A download interface for a single file.
+
+  Ranges to download are requested in get_range. Data is streamed to a
+  user-provided download_stream.
+  """
+
+  __metaclass__ = abc.ABCMeta
+
+  @abc.abstractproperty
+  def size(self):
+    """Size of the file to download"""
+
+  @abc.abstractmethod
+  def start(self, download_stream, buffer_size):
+    """Initialize downloader.
```

Comment:
<!--new_thread; commit:970afe0523d6d5673489a70e801be6553e0ffc33; resolved:0-->
Can you detail that this needs to be called before get_range calls?

See also the comments below regarding whether `download_stream` should be completely managed by the `Downloader` as opposed to currently, where it is owned by the `BufferedReader`.  In such a case, we may not need this `start` method since then the `download_stream` would be an internal detail we can deal with in the particular constructor.

Reply:
Thanks. I've removed the start() method entirely.